source/slang/slang-constructor-utility.cpp

```diff
@@ -0,0 +1,264 @@
+    {
+        if (!DiagnoseIsAllowedInitExpr(associatedDecl, nullptr))
+            return false;
+
+        // Find struct and modifiers associated with varDecl
+        StructDecl* structDecl = nullptr;
+        if (auto declRefType = as<DeclRefType>(varDeclType))
+        {
+            if (auto genericAppRefDecl = as<GenericAppDeclRef>(declRefType->getDeclRefBase()))
+            {
+                auto baseGenericRefType = genericAppRefDecl->getBase()->getDecl();
+                if (auto baseTypeStruct = as<StructDecl>(baseGenericRefType))
+                {
+                    structDecl = baseTypeStruct;
+                }
+                else if (auto genericDecl = as<GenericDecl>(baseGenericRefType))
+                {
+                    if (auto innerTypeStruct = as<StructDecl>(genericDecl->inner))
+                        structDecl = innerTypeStruct;
+                }
+            }
+            else
+            {
+                structDecl = as<StructDecl>(declRefType->getDeclRef().getDecl());
+            }
+        }
+        if (structDecl)
+        {
+            // find if a type is non-copyable
+            if (structDecl->findModifier<NonCopyableTypeAttribute>())
+                return false;
+        }
+
+        return true;
+    }
+
+    Expr* constructDefaultInitExprForVar(SemanticsVisitor* visitor, TypeExp varDeclType, VarDeclBase* decl)
+    {
+        if (!varDeclType || !varDeclType.type)
+            return nullptr;
+
+        if (!isDefaultInitializable(varDeclType.type, decl))
+            return nullptr;
+
+        ConstructorDecl* defaultCtor = nullptr;
+        auto declRefType = as<DeclRefType>(varDeclType.type);
+        if (declRefType)
+        {
+            if (auto structDecl = as<StructDecl>(declRefType->getDeclRef().getDecl()))
+            {
+                defaultCtor = _getDefaultCtor(structDecl);
+            }
+        }
+
+        if (defaultCtor)
+        {
+            auto* invoke = visitor->getASTBuilder()->create<InvokeExpr>();
+            auto member = visitor->getASTBuilder()->getMemberDeclRef(declRefType->getDeclRef(), defaultCtor);
+            invoke->functionExpr = visitor->ConstructDeclRefExpr(member, nullptr, defaultCtor->getName(), defaultCtor->loc, nullptr);
+            invoke->type = varDeclType.type;
+            return invoke;
+        }
+        else
+        {
+            return createDefaultConstructExprForType(visitor->getASTBuilder(), QualType(varDeclType.type), {});
+        }
+    }
+
+    FuncDecl* findZeroInitListFunc(StructDecl* structDecl)
+    {
+        for (auto funcDecl : structDecl->getMembersOfType<FuncDecl>())
+        {
+            if (!funcDecl->findModifier<ZeroInitModifier>())
+                continue;
+            return funcDecl;
+        }
+        return nullptr;
+    }
+
+    Expr* _constructZeroInitListFuncMakeDefaultCtor(SemanticsVisitor* visitor, StructDecl* structDecl, Type* structDeclType, ConstructorDecl* defaultCtor)
```

Comment:
this is a confusing name. I think the function should be named `constructDefaultCtorInvocationExpr`

Reply:
addressed here:
https://github.com/shader-slang/slang/pull/5107/commits/54014949789e700a589f5eea098443f3e2bc337c#diff-dad759cd0132d623bee9253550a0bce736d087bdc3157dd1e11e414cb5f6bc48R158